components/normalizer/src/lib.rs

```diff
@@ -1517,27 +1540,43 @@ macro_rules! normalizer_methods {
         /// ✨ *Enabled with the `utf8_iter` Cargo feature.*
         #[cfg(feature = "utf8_iter")]
         pub fn normalize_utf8<'a>(&self, text: &'a [u8]) -> Cow<'a, str> {
-            let up_to = self.is_normalized_utf8_up_to(text);
-            if up_to == text.len() {
-                // SAFETY: The normalization check also checks for
-                // UTF-8 well-formedness.
-                return Cow::Borrowed(unsafe { core::str::from_utf8_unchecked(text) });
+            let (head, tail) = self.split_normalized_utf8(text);
+            if tail.is_empty() {
+                return Cow::Borrowed(head);
             }
             let mut ret = String::new();
             ret.reserve(text.len());
-            let (head, tail) = text.split_at(up_to);
-            // SAFETY: The normalization check also checks for
-            // UTF-8 well-formedness.
-            ret.push_str(unsafe { core::str::from_utf8_unchecked(head) });
+            ret.push_str(head);
             let _ = self.normalize_utf8_to(tail, &mut ret);
             Cow::Owned(ret)
         }
 
+        /// Split a slice of potentially-invalid UTF-8 into maximum normalized (and valid)
+        /// prefix and unnormalized suffix such that the concatenation of the prefix and the
+        /// normalization of the suffix is the normalization of the whole input.
+        ///
+        /// ✨ *Enabled with the `utf8_iter` Cargo feature.*
+        #[cfg(feature = "utf8_iter")]
+        pub fn split_normalized_utf8<'a>(&self, text: &'a [u8]) -> (&'a str, &'a [u8]) {
+            let up_to = self.is_normalized_utf8_up_to(text);
+            if up_to <= text.len() {
+                // not using split_at_checked due to MSRV
```

Comment:
note: Typically we try and mention what Rust version is required, so that when we have the new MSRV it's easy to migrate.